server/auth/middleware.js

```diff
@@ -5,6 +5,8 @@ const isAuthenticated = (req, res, next) => {
     // ensure authenticated user is active
     if (!req.user.isActive) {
       res.status(HttpStatus.FORBIDDEN).end();
+    } else if (!req.session.twoFactor) {
+      res.status(HttpStatus.UNAUTHORIZED).end();
```

Comment:
This session twoFactor check also needs to happen in isSuperUser and isAdminUser otherwise routes protected by these will be able to bypass the 2fa.

Reply:
Hey Francis, great point. Why I didn't add them to the isSuperUser/isAdminUser middleware is this section of code in app.js. Based off my understanding, it seems that everything goes through isAuthenticated. Let me know if I am understanding this incorrectly.

app.get('/*', isAuthenticated, (req, res) => {
  let data = fs.readFileSync(client('build', 'index.html')).toString('utf8');
  data = data.replace(/window\.env\.([^ =]+)[^,;<]+/g, (match, p1) => `window.env.${p1} = '${process.env[p1] ?? ''}'`);
  res.send(data);
});

Reply:
That code in app.js is the "catch-all" handler. Above that, where require('./routes') is called, is where all the specific routes implemented in the files in the routes folder are loaded, and the catch-all is only used when an incoming request doesn't match any of the previous routes. The catch-all is used to serve up the base index.html for the client SPA. If you look in routes/api/users.js you'll see some routes that use middleware.isAdminUser instead of isAuthenticated.

Reply:
That makes perfect sense. A silly mistake from me as I forgot that routing goes from top of the page to the bottom of the page. I am aligned here!